usaspending_api/etl/elasticsearch_loader_helpers/controller.py

```diff
@@ -189,7 +190,7 @@ def extract_transform_load(task: TaskSpec) -> None:
             msg = f"Partition #{task.partition_number} failed after an error was previously encountered"
             logger.warning(format_log(msg, name=task.name))
         else:
-            logger.error(format_log(f"{task.name} failed!", name=task.name))
+            logger.exception(format_log(f"{task.name} failed!", name=task.name))
```

Comment:
Here's the full snippet
```python
    try:
        records = task.transform_func(task, extract_records(task))
        if abort.is_set():
            f"Prematurely ending partition #{task.partition_number} due to error in another process"
            logger.warning(format_log(msg, name=task.name))
            return
        if len(records) > 0:
            success, fail = load_data(task, records, client)
        else:
            logger.info(format_log("No records to index", name=task.name))
            success, fail = 0, 0
        with total_doc_success.get_lock():
            total_doc_success.value += success
        with total_doc_fail.get_lock():
            total_doc_fail.value += fail
    except Exception:
        if abort.is_set():
            msg = f"Partition #{task.partition_number} failed after an error was previously encountered"
            logger.warning(format_log(msg, name=task.name))
        else:
            logger.exception(format_log(f"{task.name} failed!", name=task.name))
            abort.set()
```

I think at most it will record the exception twice. Since once it logs the exception, it sets `abort.set()` and the caught exception will just log a warning if it enters there again for a different process, by finding that `if abort.is_set():` is `True`. 

So that has one of two outcomes: 
1. The worker process already caught and logged this exception, and then raised it up to the controller process, where it will be logged a second time. 
1. It was an _unhandled_ exception (e.g. happening somewhere that was not expected, and therefore had not `try-except` block wrapping it) and so it bubbles up to the controller, where it is logged for the first and only time. 

This was added more for the latter scenario to make sure that we log exceptions that were not expected, rather than just logging that an error occurred, but not seeing the traceback.

To do further cleanup what we _could_ do is always just log an ERROR in subprocesses, and ensure that this controller `try-except` logs the actual exception. There are cases where I saw `SystemExit` though so i'm not sure it will always make it up to the top controller process. For now might just be best to err on the side of caution and easiest to ride with the 2x redundancy and revisit if the logs get really painful to read.

Reply:
I think it is possible to have dozens of processes simultaneously fail due to a cascading issue. And as far as I can tell by observed behavior and the docs, it is possible for a number of processes to check `abort.is_set()` before the first process can set it using `abort.set()`. I think the logging statements can slow down the execution enough and I don't think the call to set actually obtains a lock.

Feel free to read the notes yourself: https://docs.python.org/3.7/library/threading.html#threading.Event

I assume that most of these unfortunate scenarios will occur locally, so I won't drag the PR over this.